### ISSUE TITLE
Android build causes compile error with `#include <torch/torch.h>` from v1.12.0 because PadNd.h is not found

### ISSUE THREAD

**Issue by xuzijian629** [pytorch/pytorch#81198]
### 🐛 Describe the bug

# Problem
Android build (and maybe other mobile builds also) has incomplete dependency which causes compile error just by `#include <torch/torch.h>`.

By building a C++ project with the mobile library, you will get an error like
```cpp
/home/joe/android-build/third_party/libtorch-1.12.0-android/include/torch/csrc/api/include/torch/nn/functional/padding.h:4:10: fatal error: 'ATen/native/PadNd.h' file not found
#include <ATen/native/PadNd.h>
         ^~~~~~~~~~~~~~~~~~~~~
1 error generated.
...
In file included from /home/joe/android-build/third_party/libtorch-1.12.0-android/include/torch/csrc/api/include/torch/torch.h:3:
In file included from /home/joe/android-build/third_party/libtorch-1.12.0-android/include/torch/csrc/api/include/torch/all.h:13:
In file included from /home/joe/android-build/third_party/libtorch-1.12.0-android/include/torch/csrc/api/include/torch/nn.h:4:
In file included from /home/joe/android-build/third_party/libtorch-1.12.0-android/include/torch/csrc/api/include/torch/nn/functional.h:11:
In file included from /home/joe/android-build/third_party/libtorch-1.12.0-android/include/torch/csrc/api/include/torch/nn/functional/normalization.h:4:
```

# Reproduction
```bash
$ git clone https://github.com/pytorch/pytorch.git -b v1.12.0 --depth 1 --recursive --shallow-submodules -j4
$ wget https://dl.google.com/android/repository/android-ndk-r21-linux-x86_64.zip -q
$ unzip -q /home/joe/android-build/third_party/android-ndk-r21-linux-x86_64.zip
$ ANDROID_ABI="arm64-v8a" ANDROID_NDK="$(pwd)/android-ndk-r21" ./scripts/build_android.sh
```

Then you will get the compiled libraries
```bash
$ ls ./build_android/install
include/  lib/  share/
```

Check `ATen/native/PadNd.h` will be included by `#include <torch/nn/functional.h>` (that will be included by `#include <torch/torch.h>`).
```
$ grep PadNd.h ./build_android/install/include/torch/csrc/api/include/torch/nn/functional/padding.h
#include <ATen/native/PadNd.h>
```

But you don't have `PadNd.h` in your build library
```bash
$ find ./build_android/install -name PadNd.h
(no output)
```

# Cause of the error
For mobile builds, `ATen/native/*.h` headers are removed from `INSTALL_HEADERS`. https://github.com/pytorch/pytorch/blob/v1.12.0/aten/src/ATen/CMakeLists.txt#L528-L532

However, `ATen/native/PadNd.h`, was introduced from in #73431 (appeared in v1.12.0 release), and is required in [torch.nn.functional/padding.h](https://github.com/pytorch/pytorch/blob/v1.12.0/torch/csrc/api/include/torch/nn/functional/padding.h#L4) which will be loaded by
```cpp
#include <torch/torch.h>
```
even in mobile builds.

### Versions

v1.12.0

For android versions, please see above **Reproduction** section.

cc @ezyang @gchanan @zou3519

**Comment by Olivia-liu** [comment]
@iseeyuan Hi Martin! This is related to ATen. Do you mind taking a look? Thank you!

**Comment by cccclai** [comment]
Just want to make sure the environment is set up correctly. Are you able to build pytorch from source? (https://github.com/pytorch/pytorch#from-source)

**Comment by cccclai** [comment]
Just want to make sure the environment is set up correctly. Are you able to build pytorch from source? (https://github.com/pytorch/pytorch#from-source). Also this is the command I used to build for Android:
```
./scripts/build_pytorch_android.sh x86
```

**Comment by xuzijian629** [comment]
> Just want to make sure the environment is set up correctly.

Yes, my command is as written in the **Reproduction** section:
```
 ANDROID_ABI="arm64-v8a" ANDROID_NDK="$(pwd)/android-ndk-r21" ./scripts/build_android.sh
```

But I think the error is caused because `PadNd.h` is a newly created file and not properly included in CmakeLists.txt (or, PadNd.h should be placed in a different location)

**Comment by dreiss** [comment]
@ezyang , I believe we decided that to support selective build, we'd have no direct dependencies from torch to aten/native.  Is that still right?  I think this one sneaked past the internal build because the reference is only in a header, and we're far less strict with headers than cpp dependencies.  Any idea where we could put PadNd.h so it's accessible from both aten/native and torch/csrc/api?

**Comment by ezyang** [comment]
pretty normal spot would be aten/ in that case.

**Comment by gorkemalgan** [comment]
I am facing the same issue on ios as well. I tried both pre-built libtorch from Cocoapods and built it from source by build_ios.sh. Both resulted the same error

**Comment by cccclai** [comment]
Acutally I meant 

> > Just want to make sure the environment is set up correctly.
> 
> Yes, my command is as written in the **Reproduction** section:
> 
> ```
>  ANDROID_ABI="arm64-v8a" ANDROID_NDK="$(pwd)/android-ndk-r21" ./scripts/build_android.sh
> ```
> 
> But I think the error is caused because `PadNd.h` is a newly created file and not properly included in CmakeLists.txt (or, PadNd.h should be placed in a different location)

Actually to the command to build PyTorch from source (not for Android) is 
```
MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py install
```
If it works, likely you've gotten all the required dependency. 

Also the command I usually use to build PyTorch for Android is
```
./scripts/build_pytorch_android.sh x86
```

**Comment by cccclai** [comment]
> I am facing the same issue on ios as well. I tried both pre-built libtorch from Cocoapods and built it from source by build_ios.sh. Both resulted the same error

Does it mean the prebuilt one works for you but run into the same issue when building PyTorch for iOS from scratch? Is `build_ios.sh` the command line you use?

**Comment by gorkemalgan** [comment]
No, I face the same issue if I also use the prebuilt ones from Cocoapods. Yes, I used build_ios.sh for building from source.

**Comment by TimPushkin** [comment]
Still having this issue when using prebuild PyTorch 1.12.2 on Android (unpacked from https://repo1.maven.org/maven2/org/pytorch/pytorch_android/1.12.2/pytorch_android-1.12.2.aar). Seems like the PR meant to fix this wasn't merged. Any updates?

**Comment by TimPushkin** [comment]
It would be also great to know if just manually copying `PadNd.h` into the headers of the unpacked AAR will allow to temporarily bypass the issue? I mean, it makes the Android app I'm developing build but won't anything break if I actually call something that uses  `PadNd.cpp`?

**Comment by ezyang** [comment]
Copying it in should be ok
